cmd/influxd/upgrade/security.go

```diff
@@ -1,11 +1,283 @@
 package upgrade
 
+// Security upgrade implementation.
+// Generates script for upgrading 1.x users into tokens in 2.x.
+
 import (
 	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"sort"
+	"strings"
+	"text/template"
+
+	"github.com/influxdata/influxdb/v2/v1/services/meta"
+	"github.com/influxdata/influxql"
 	"go.uber.org/zap"
 )
 
-// generateSecurityScript generates security upgrade script.
-func generateSecurityScript(v1 *influxDBv1, dbBuckets map[string][]string, log *zap.Logger) error {
-	return errors.New("not implemented")
+// generateSecurityScript creates security upgrade script with 1.x users mapped to 2.x tokens.
+func generateSecurityScript(v1 *influxDBv1, targetOptions optionsV2, dbBuckets map[string][]string, log *zap.Logger) error {
+	// check if there any 1.x users at all
+	v1meta := v1.meta
+	if len(v1meta.Users()) == 0 {
+		log.Warn("There are no users in 1.x, no script will be created.")
+		return nil
+	}
+
+	// create helper
+	helper := &securityScriptHelper{
+		log: log,
+	}
+	if err := helper.init(); err != nil {
+		return err
+	}
+
+	// check if target buckets exists in 2.x
+	proceed := helper.checkDbBuckets(v1meta, dbBuckets)
+	if !proceed {
+		return errors.New("upgrade: there were errors/warnings, please fix them and run the command again")
+	}
+
+	// create output
+	var output *os.File
+	var isFileOutput bool
+	if targetOptions.securityScriptPath == "" {
+		output = os.Stdout
+	} else {
+		file, err := os.OpenFile(targetOptions.securityScriptPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0777)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+		output = file
+		isFileOutput = true
+	}
+
+	// get `influx` absolute path
+	exePath, err := os.Executable()
+	if err != nil {
+		return err
+	}
+	influxExe := filepath.Join(filepath.Dir(exePath), "influx")
+
+	// template data type
+	type dataObject struct {
+		AdminUsers   []string
+		IgnoreUsers  []string
+		UpgradeUsers []string
+		UserArgs     map[string][]string
+		Exe          string
+		TargetAdmin  string
+		TargetOrg    string
+		TargetToken  string
+		IsFileOutput bool
+		InfluxExe    string
+	}
+
+	// data
+	data := dataObject{
+		UserArgs:     make(map[string][]string),
+		TargetAdmin:  targetOptions.userName,
+		TargetOrg:    targetOptions.orgName,
+		TargetToken:  targetOptions.token,
+		IsFileOutput: isFileOutput,
+		InfluxExe:    influxExe,
+	}
+
+	// fill data with users and their permissions
+	for _, row := range helper.sortUserInfo(v1meta.Users()) {
```

Comment:
This is good that the user data is sorted, to ensure consistent ordering in the script output